llama-cpp.nix

```diff
@@ -0,0 +1,118 @@
+{ lib
+, stdenv
+, cmake
+, ninja
+, pkg-config
+, symlinkJoin
+, python3
+, openmpi
+, openblas
+, cudaPackages
+, rocmPackages
+, clblast
+, Accelerate ? null
+, MetalKit ? null
+, CoreVideo ? null
+, CoreGraphics ? null
+, useOpenCL ? false
+, useCuda ? false
+, useRocm ? false
+}:
+
+let
+  inherit (lib) optional optionals versionOlder;
+  isDefault = !useOpenCL && !useCuda && !useRocm;
+
+  # Give a little description difference between the flavors.
+  descriptionSuffix = if useOpenCL then
+    " (OpenCL accelerated)"
+  else if useCuda then
+    " (CUDA accelerated)"
+  else if useRocm then
+    " (ROCm accelerated)"
+  else if (MetalKit != null) then
+    " (MetalKit accelerated)"
+  else "";
+
+  llama-python = python3.withPackages (ps: [ ps.numpy ps.sentencepiece ]);
+
+  # See ./overlay.nix for where these dependencies are passed in.
+  defaultBuildInputs = builtins.filter (p: p != null) [
+    Accelerate
+    MetalKit
+    CoreVideo
+    CoreGraphics
+  ];
+
+  cudaBuildInput = symlinkJoin (with cudaPackages; {
```

Comment:
IIRC llama.cpp uses `find_package(CUDAToolkit)`. This means you don't have to use `symlinkJoin` nor `cudaPackages.cudatoolkit`. Both are expensive and discouraged (mostly because of the huge runtime closure).

Cf. e.g. pytorch from Nixpkgs:

https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L360-L361
https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L365-L367
https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L303
https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L333-L356

Reply:
Oh, I was hoping github would render the previews

Reply:
Great. I took the suggestion and added the specific build and runtime dependencies needed. It's much cleaner now.